### ISSUE TITLE
Create inherited views

### ISSUE THREAD

**Issue by kaerdsar** [OCA/web#705]
I would like to have a feature in Odoo where a change in the arch of a view can be saved as a new inherited view. This feature can be useful to handle the version history in Odoo views like website pages and reports.

This issue has the goal to discuss some of the approaches we can have in order to create a good solution. Recently I read a paper suggested by @hbrunn, about XML change detections, which I think is a good start point. I created an example module using an C++ implementation of the algorithm in order to do a proof of concept. It is not completely working, currently is only to test the idea, it can be checked here: https://github.com/kaerdsar/odoo-addons/tree/master/web_view_history.

Some points about the example:
- The algorithm in the paper does not take care about position of element siblings.
- It would be better to have a python library instead of the C++ implementation. In this case, we can avoid to work with files and get a list of changes instead of a new XML with processing instructions.
- We should also try to add a feature to show the changes by detecting the inherited views created with the module. I think it would be very useful to check the changes, accept or revert them.

@pedrobaeza @dreispt It would be great to get some feedbacks about this topic.

**Comment by dreispt** [comment]
That would be a great addition!
IMO it should be an improvement to the [module_prototyper](https://github.com/OCA/server-tools/tree/10.0/module_prototyper) module.

I would very much prefer to see a Python implementation: this will run at "design time", so performance is not so important.

**Comment by hbrunn** [comment]
why don't you write this in C and [wrap it in lxml](http://lxml.de/capi.html)? doesn't look so complicated

**Comment by kaerdsar** [comment]
Interesting point, I will try later to wrap the library with Cython.

Other point. Imagine this scenario:
- I have a website page.
- I edit the page in the website editor and after save a new inherited view is created.
- When the page is rendered the inherited view is shown.
- I try to edit again on the same paragraph.
- The module will try to create an inherited view based on another inherited view. This has high risk that XPATH works with non existing elements and get error.

How should we handle news inherited views when the view already has at least one inherited view?

**Comment by sbidoul** [comment]
@kaerdsar interesting work. Have you looked at pure python implementations of the xml diff algorithm? eg https://pypi.python.org/pypi/xmldiff or https://github.com/cfpb/xtdiff ?

**Comment by kaerdsar** [comment]
@sbidoul thanks for joining. I already tested xtdiff, it was the first one I checked. I did an example with 2 large XML and the time was not good, it took a lot of time to  return the result.

I  haven't check xmldiff, I'll do it. Thanks

**Comment by dreispt** [comment]
AFAICT for the particular use case of views, the performance is not that critical: the XML docs are relatively small, and the diff process runs at design time, not run time.

**Comment by kaerdsar** [comment]
I haven't had enough time to check well the libraries in order to decide wether we should use the xmldiff python implementation or create a wrap in lxml for the C++ implementation. For testing purposes, I updated the C++ library to use standard input/output instead of files. 

At the moment I was trying to focus in the view history concept based on inherited views. I updated the code sample posted in the first comment. The main idea is:
- For the first change, generate a new inherited view based on the changes.
- For next changes, set existing inherited view to active False and create a new one with all the changes based on the original view.
